### PR TITLE
NO MERGE: A try to update the return type of mostSpecificRouteConfig

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by wbpcode** [comment]
A quick try to update the return type of `mostSpecificPerFitlerConfig`.

Here is some results:
- There a ton of changes. Almost all http filters need to be updated to avoid any possible dangling ref/ptr. This PR havn't fix the tests.
- I cannot ensure my patch is safe. I am not expert to every filter. I cannot ensure my patch could solve the lifetime problem compeletely. Some filters have filter-specific logic to handle the config, this may introduce more uncertainty.
- Maybe always keep a shared pointer in the route will be easiest way. But I think it's a little heavy. And It denpends on filter's implementation. For example, lua will get the per route config in the `decodeHeaders` phase and use it in the whole decoding phase. And it will get the per route config again in the `encodeHeaders` phase and use it in the whole encoding phase. So, we need to get and store the route in both `decodeHeaders` phase and `encodeHeaders` phase.
- Seems that exposing these lifetime problems to filters is more complex than our initial thought.


Back to the #26045, maybe keep a trash can in the ActiveStream is an possible alternative compromise.

cc @mattklein123

**Comment by mattklein123** [comment]
I think we should clean this up. I'm not sure I understand the complexity. Any time anyone is storing a raw pointer to a route can't we just store a shared_ptr?

**Comment by wbpcode** [comment]
In fact almost no filter store a raw pointer of route. The route specific filter config bring the most complexity.

There are lots filter-specific logic about the config usage make we must do lots change to get a clean code.

**Comment by mattklein123** [comment]
Can you share an example of what is complicated? I'm probably just missing something. It seems like we are either storing a raw pointer or a shared pointer?

File: envoy/http/filter.h [hunk 1]
```diff
@@ -399,7 +399,7 @@ class StreamFilterCallbacks {
    * This is a helper to get the route's per-filter config if it exists, otherwise the virtual
    * host's. Or nullptr if none of them exist.
    */
-  virtual const Router::RouteSpecificFilterConfig* mostSpecificPerFilterConfig() const PURE;
```
Comment (mattklein123) [inline envoy/http/filter.h:402]: Why don't we just return a shared_ptr here directly?

**Comment by wbpcode** [comment]
> Can you share an example of what is complicated? I'm probably just missing something. It seems like we are either storing a raw pointer or a shared pointer?

That's the essence, but the practice is a little more complicated.

Some filters may extract a sub-config from the route specific route config and then start a async operation directly. In the origin code, there are no any 'store'. But now we always need add a store logic to store a shared pointer into the filter instance regardless of where we do the config extraction to ensure the lifetime safety.

I don't mean we can't achieve it, but more code change are necessary than our initial thought.

This make me think another questions:
- 1. Will this way have better performance? The answer should be no. Because more shared pointer is used. Although it shouldn't be worse because the atomic operation of shared pointer is super fast.
- 2. Will the code more safe? I think the answer should be no? The safety shouldn't have any change.
- 3. Will this be more friendly to developers? I think the answer still be no? Now every filter need care about the lifetime of route-related config.

File: envoy/http/filter.h (hunk 1 above)
Comment (wbpcode) [reply envoy/http/filter.h:402]: Almost no filter use this method directly. most filters use a utility function to get the typed config. So I create a variant of origin utility function to get a typed shared pointer. For the filters that need not keep the ownership of config, they can use origin utility function directly.

**Comment by mattklein123** [comment]
> This make me think another questions

You are correct, but it does leave some very confusing technical debt in the HCM which is already very complicated. I still feel we should try to clean this up but happy to be outvoted on this if you want to leave a TODO for later. If that's the case, please awdd some detailed comments/TODO around why we are using that delayed storage list and how we might clean it up in the future.

**Comment by wbpcode** [comment]
A tentative conclusion is keep current implementation in the #26045. Some detailed comments and TODO will be added.

**Pull request closed without merging**